src/symbol/collision_index.js

```diff
@@ -143,99 +124,123 @@ class CollisionIndex {
+            if (labelToScreenMatrix) {
+                const screenSpacePath = projectedPath.map(p => projection.project(p, labelToScreenMatrix));
+
+                // Do not try to place collision circles if even of the points is behind the camera.
+                // This is a plausible scenario with big camera pitch angles
+                if (screenSpacePath.some(point => point.signedDistanceFromCamera <= 0))
+                    projectedPath = [];
+                else
+                    projectedPath = screenSpacePath.map(p => p.point);
             }
 
-            const projectedPoint = this.projectPoint(posMatrix, anchorPointX, anchorPointY);
-            const radius = tileUnitRadius * tileToViewport;
-
-            const atLeastOneCirclePlaced = placedCollisionCircles.length > 0;
-            if (atLeastOneCirclePlaced) {
-                const dx = projectedPoint.x - placedCollisionCircles[placedCollisionCircles.length - 4];
-                const dy = projectedPoint.y - placedCollisionCircles[placedCollisionCircles.length - 3];
-                // The circle edges touch when the distance between their centers is 2x the radius
-                // When the distance is 1x the radius, they're doubled up, and we could remove
-                // every other circle while keeping them all in touch.
-                // We actually start removing circles when the distance is √2x the radius:
-                //  thinning the number of circles as much as possible is a major performance win,
-                //  and the small gaps introduced don't make a very noticeable difference.
-                const placedTooDensely = radius * radius * 2 > dx * dx + dy * dy;
-                if (placedTooDensely) {
-                    const atLeastOneMoreCircle = (k + 8) < collisionCircles.length;
-                    if (atLeastOneMoreCircle) {
-                        const nextBoxDistanceToAnchor = collisionCircles[k + 8];
-                        if ((nextBoxDistanceToAnchor > -firstTileDistance) &&
-                        (nextBoxDistanceToAnchor < lastTileDistance)) {
-                            // Hide significantly overlapping circles, unless this is the last one we can
-                            // use, in which case we want to keep it in place even if it's tightly packed
-                            // with the one before it.
-                            markCollisionCircleUsed(collisionCircles, k, false);
-                            continue;
-                        }
-                    }
+            let segments = [];
+
+            if (projectedPath.length > 0) {
+                // Quickly check if the path is fully inside or outside of the padded collision region.
+                // For overlapping paths we'll only create collision circles for the visible segments
+                const minPoint = projectedPath[0].clone();
+                const maxPoint = projectedPath[1].clone();
```

Comment:
Is it impossible for `projectedPath[1]` to be undefined?

Reply:
It should be impossible. `placeGlyphAlongLine` constructs a path from the anchor point to the position of first/last symbol. Projected path is constructed from anchor-to-first and anchor-to-last segments. I added an extra assertion.